accepted/2021/experimental-features/experimental-features.md

```diff
@@ -0,0 +1,390 @@
+  which parts aren't.
+* Customer have an easy way to audit if their applications or libraries are
+  using experimental features.
+* Customers can't accidentally depend on experimental features, either directly
+  or indirectly through some library.
+* We can deliver a reasonable experience even when an experimental feature can't
+  be supported in all the workloads, for example, by failing with a sensible
+  error message.
+* We have a unified experience for turning on experimental features, no matter
+  whether they are in a runtime, a compiler, a library, or span multiple parts
+  of the system.
+* We don't have to do unnatural acts to add an experimental feature, by, for
+  example, having to build a plug-in based system that allows swapping in and
+  out files.
+
+### Non-Goals
+
+* Allowing Microsoft to turn features on or off remotely, e.g. A/B testing
+* Designing a mechanism for shipping experimental features as separate packages
+  or installers
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# compiler and IDE teams
+* Library teams
+* Project system team
+* SDK team
+* MSBuild team
+* NuGet team
+
+## Design
+
+### Property in project files
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <EnableExperimentalFeature Include="StaticInterfaceMembers" />
+  </ItemGroup>
+
+</Project>
+```
+
+### Assembly info generation
+
+When we generate the `AssemblyInfo.cs`, we'll include all experimental features
+as an assembly level attribute (the attribute and the meaning is discussed
+[later in this document](#api-analyzer)).
+
+For example, this project information:
+
+```xml
+<Project>
+
+  <ItemGroup>
+    <EnableExperimentalFeature Include="StaticInterfaceMembers" />
+    <EnableExperimentalFeature Include="FancyGC" />
+  </ItemGroup>
+
+</Project>
+```
+
+Would result in these assembly-level attributes:
+
+```C#
+[assembly: ExperimentalFeature("StaticInterfaceMembers")]
+[assembly: ExperimentalFeature("FancyGC")]
+```
+
+This means you can look at a binary and tell which experimental features were
+used to build it. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+The `ExperimentalFeature` should be written into the `runtime.json` file such
+that the host/runtime can use it to make decisions.
+
+### Compilation context
+
+The `ExperimentalFeature` needs to be passed to the compilers. Roslyn (C#/VB)
+already has a mechanism for this which is exposed to both the compiler as well
+as for analyzers/source generators. Alternatively, we could say the information
+about experimental features can be retrieved from the generated assembly-level
+attributes.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
```

Comment:
This will protect compile time. Do we also need to protect runtime time, like reflection exposing the experimental surface, and apps and libraries taking accidental dependency on it?

Reply:
This includes indirect dependencies on reflection, like Powershell. Should Powershell be invited to this conversation to provide feedback? It may require work on their side.